doc/source/DBEP/DBEP02.rst

```diff
@@ -0,0 +1,167 @@
+===================================================
+ Promote Resource / Datum to first-class documents
+===================================================
+
+.. contents::
+   :local:
+
+
+Status
+======
+
+**Discussion**
+
+
+Branches and Pull requests
+==========================
+
+
+
+Abstract
+========
+
+Currently *Resource* and *Datum* are directly inserted into the
+*AssetRegistry* by *ophyd*.  This breaks the document abstractions
+by making a specific consumer 'special'.
+
+
+
+Detailed description
+====================
+
+An odd asymmetry in how databroker works is that the documents for
+*HeaderSource* and *EventSource* are emitted by the *RunEngine* and
+can be subscribed to by one or more consumers.  Each consumer is
+notionally independent, each receive all of the documents, and do not
+need to coordinate in any way (or even be aware of each others
+existence). In contrast, the *Resource* and *Datum* documents are
+inserted directly into an *AssetRegistry* by the *ophyd* objects.
+This breaks the separation we have between the data collection process
+/ hardware, the generation of the documents, and the consumers of
+those documents and leads to several unfortunate situations:
+
+ - *ohpyd* objects hold an instance of an *AssetRegisty*
+ - we need to keep track of _which_ *AssertRegistry* things were
+   inserted into
+ - consumers that want access to the asset documents need to also have
+   a handle to the database that the objects are inserting into
+
+The proposed solution is to promote *Resource* and *Datum* documents
+to be peer documents with *Start*, *Stop*, *Descriptor* and *Event*.
+They will appear in the document stream and be inserted into
+*DataBroker* via ``db.insert``.  This eliminates the 'special'
+side-band communication and brings all consumers back to the same
+footing.  This will require coordinated changes to *event-model*,
+*databroker*, *bluesky*, and *ophyd*.
+
+
+
+Implementation
+==============
+
+Currently, *ophyd* is responsible for collecting all of the values for
+the *Resource* and *Datum* documents except for the uids.  The uids
+are generated by calls to ``reg.register_*`` and the datum uids are
+subsequently returned to the *RunEngine* via ``obj.read``.  The
+proposed change is:
+
+ 1. *ophyd* objects would be responsible for generating the full *Resource*
+    and *Datum* documents and providing them to the *RunEngine* to be
+    emitted (likely should provide some helpers).
+
+    a. Similar to the current documents, a *Resource* must be emitted
+       before any *Datum* that refers to it per *Start* / *Stop* pair
+    b. an identical (including uid) *Resource* maybe emitted more than
+       once, the consumers will need to handle this.
+    c. an identical *Datum* document may be returned the *RunEngine*
```

Comment:
My answer to the question here: We could attempt to de-duplicate within a given plan, but we can't de-duplicate globally without caching an unbounded set of UUIDs. Since consumers are potentially long-running and don't know the scope of a plan (they only see runs) I don't it's useful to attempt de-duplication at all.

Reply:
I would also be tempted to leave them as is. Duplicate `datums` sounds like a bug to me and something that shouldn't happen. So if it did (and it will) having traces of it at the document level may be easier to debug than searching around at the ophyd level.
(Correct me if I misunderstood. I understood this as "should we curate/clean documents as they the `RE` receives them?" I would say no)

Reply:
The issue I am thinking of here is if we have a device that returns from read things that look like

```python
{'image': UID1,
 'background': UID2,
 'flatfield': UID3}
```

where the 'background' and 'flaitfield' may stay constant of many calls to `read`.  Updated this with some requirements on the latch logic around this.